File: Mergesort/mergesort.scala
def merge(nums1 : List[Int], nums2 : List[Int]) : List[Int] = {
	(nums1, nums2) match {
		case (Nil, nums2) => nums2
		case (nums1, Nil) => nums1
		case (x::xs, y::ys) => {
			if (x < y) x :: merge(xs, nums2)
			else y :: merge(nums1, ys)
		}
	}
}

def merge_sort(nums : List[Int]) : List[Int] = {
	if (nums.length <= 1) nums
	else {
		val s = nums.length / 2
		merge(merge_sort(nums.take(s)), merge_sort(nums.drop(s)))
	}
}



println(merge_sort(List(3, 6, 1, 9, 3, 6, 8, 2)))

File: Mergesort/mergesort.js
var js_test = require('./../Test/JavascriptTest')


function mergesort(nums) {

	const merge = (xs, ys) => {
		if (xs.length == 0) {
			return ys
		} else if (ys.length == 0) {
			return xs
		}
		else {
			if (xs[0] < ys[0]) {
				return [xs[0]].concat(merge(xs.slice(1, xs.length), ys))
			} else {
				return [ys[0]].concat(merge(xs, ys.slice(1, ys.length)))
			}
		}
	}

	if (nums.length < 2) {
		return nums
	}

	let mid = parseInt(nums.length / 2)
	return merge(mergesort(nums.slice(0, mid)), mergesort(nums.slice(mid, nums.length)))
}


// test
console.log(js_test.test(mergesort))
